Tercer entregable

JaimeGarcia-Jimmy opened the pull request:
-Capa datasource para books
-Configuracion de depenedencias para poder usar los comandos ktlint y ktllintformat
-Correccion de bug en toolbar

JaimeGarcia-Jimmy committed:
-Capa datasource para books
-Configuracion de depenedencias para poder usar los comandos ktlint y ktllintformat
-Correccion de bug en toolbar

JaimeGarcia-Jimmy committed:
-Capa datasource para bookOrders
-Interfaces para capa de repositories
-Reformatear petcion al endpoint  https://api.bitso.com/v3/ticker para usar RxJava
-Se elimino la clase ApiClient

JaimeGarcia-Jimmy committed:
-Pruebas unitarias para AvailableBooksViewModel

DiegoMedinaP commented on app/src/main/AndroidManifest.xml:
```diff
@@ -9,6 +9,7 @@
     <application
         android:name=".CriptoApplication"
         android:allowBackup="true"
+        android:enableOnBackInvokedCallback="true"
```
para qué se utiliza esto?

JaimeGarcia-Jimmy replied:
Ya eliminé ese atributo de application. Lo había puesto porque cuando estaba implementando RxJava, me aparecía un warning que me recomendaba ponerlo. Como dice algo de callbacks, pensé que estaba relacionado. Pero solo fue coincidencia.

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/data/repositories/BookOrdersRepositoryImpl.kt:
```diff
@@ -0,0 +1,117 @@
+    override fun getBookById(bookId: String): Flow<Resource<Book>> = flow {
+        emit(Resource.loading())
+
+        val dbFlow: Flow<Resource<Book>> = bookOrdersLocalDatasource.getBookByIdFlow(bookId).map {
```
`val dbFlow: Flow<Resource<Book>> =` no es necesario realizar esta asignacion dado que no se esta ocupando en ningun lado

JaimeGarcia-Jimmy replied:
Si lo estoy emitiendo en la linea 48: emitAll(dbFlow)

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/data/repositories/BookOrdersRepositoryImpl.kt:
```diff
@@ -0,0 +1,117 @@
+        try {
+            updateBook(bookId)
+        } catch (e: Exception) {
+            emit(Resource.error("La peticion fallo", e))
```
es conveniente manejar los mensjaes de error en la carpeta de strings y obtenerlos desde ahí

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/data/repositories/BookOrdersRepositoryImpl.kt:
```diff
@@ -0,0 +1,117 @@
+        }
+
+        emitAll(dbFlow)
+    }.flowOn(Dispatchers.IO)
```
el Dispatchers.IO es un objeto que se esta creando dentro de tu repositorio, si tu quisieras testear este metodo, no podrias hacerlo en otro hilo o te costaria mucho trabajo dado que lo estas creando aqui, lo recomendable en estos casos es que el RepositoryImpl reciba como parametro el dispatcher , puedes inyectarselo con hilt y de esa forma cuando hagas testing puedes sustituir este dispatcher por uno de prueba

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/data/repositories/BookOrdersRepositoryImpl.kt:
```diff
@@ -0,0 +1,117 @@
+    }.flowOn(Dispatchers.IO)
+
+    override suspend fun updateBook(bookId: String) =
+        withContext(Dispatchers.IO) {
+            val disposable = bookOrdersRemoteDataSource.getBookByTickerObservable(bookId)
+                .subscribeOn(Schedulers.io())
+                .observeOn(Schedulers.io())
+                .subscribe(::handleGetBookByTickerSuccess, ::handleGetBookByTicketError)
+            compositeDisposable.add(disposable)
+        }
```
esta mezcla entre una funcion de suspencion y Rx se ve un poco tricky, recomiendo que la cambies por una funcion normal (sin el suspend) y que regreses el observable directamente, para que sea el viewModel el encargado de suscribirse y observar.

Del mismo modo que en el comentario de arriba si quieres ponerle el Schedulers yo te recomiendo que se lo inyectes y no crees el Schedulers.io() directamente

JaimeGarcia-Jimmy replied:
Esa función de updateBook originalmente funcionaba con corrutinas. Se me olvidó quitar las partes de suspend y withConext, cuando la modifiqué. Como no me daba error, se me pasó.
Dejé el subscribe y observe ahí en el mismo repositorio. Porque cuando la respuesta de la fuente remota es exitosa, actualizo la fuente local con esa respuesta. Me pareció lo mejor manejar esa lógica en la capa de repositorio. ¿Crees que es mas conveniente pasar esa logica al viewmodel y desde ahi ejecutar el update en la base de datos (pasando nuevamente por el repositorio)?

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/ui/AvailableBooksFragment.kt:
```diff
@@ -20,12 +18,12 @@ import kotlinx.coroutines.flow.collect
     private val binding: FragmentAvailableBooksBinding get() = _binding!!
     private val viewModel: AvailableBooksViewModel by viewModels<AvailableBooksViewModel>()
-    private val adapter = BookAdapter( mutableListOf() ){ bookId -> openBookDetail(bookId) }
+    private val adapter = BookAdapter(mutableListOf()) { bookId -> openBookDetail(bookId) }
```
te recomiendo que cambies tu `RecyclerView.Adapter` por un `ListAdapter` tiene mejor performance, es mas fluido y actualiza unicamente los elementos que se modifciaron si necesidad de elimnar y agregar nuevamente todos los elementos de la lista

DiegoMedinaP commented on app/src/main/java/com/example/criptomonedas/ui/BookDetailFragment.kt:
```diff
@@ -21,36 +21,38 @@ import kotlinx.coroutines.flow.collect
     private val viewModel: BookDetailViewModel by viewModels()
-    private val askAdapter = AskAdapter( mutableListOf() )
-    private val bidAdapter = BidAdapter( mutableListOf() )
+    private val askAdapter = AskAdapter(mutableListOf())
+    private val bidAdapter = BidAdapter(mutableListOf())
```
Igual recomiendo cambiar por un ListAdapter

DiegoMedinaP commented on app/src/test/java/com/example/criptomonedas/ui/viewmodels/AvailableBooksViewModelTest.kt:
```diff
@@ -0,0 +1,66 @@
+    }
+
+    @Test
+    fun `Loading state after executing updateBooks works`() = runBlocking {
+        whenever(booksRepository.updateBooks()).doSuspendableAnswer {
+            withContext(Dispatchers.IO) { delay(5000) }
+        }
+        viewModel = AvailableBooksViewModel(booksRepository)
+        viewModel.updateBooks()
+        Assert.assertEquals(Resource.loading<Resource<List<Book>>>(), viewModel.booksList.value)
+    }
```
cuando testeamos el viewModel lo mejor es verificar si entro a un estado o no, podrias ocupar el verify(exactly = 1) para saber que el codigo entro a ese estado una vez 

No es recomendable poner delays para forzar a mostrar un estado por varias razones:
1 existe la posibilidad que en algun momento o en ciertos casos, la respuesta caiga despues del tiempo que tu le determinaste, por lo que se volveria un flaky test (tests que solo pasan bajo ciertas condiciones o no en todas las ocaciones)

2 Imagina que fuera un aplicacion mas grande y tenemos que hacer 30 o 40 tests similares, estamos aumentando el tiempo de verificacion en 5000 milisegundos por test

JaimeGarcia-Jimmy committed:
-Inyeccion de CoroutineDispatcher y Scheduler en viewmodels y repositories.
-Se modifico el metodo updateBook en BookOrdersRepositoryImpl para que no sea una funcion suspendible, ni corra dentro de una corrutina.
-Se elimino la propiedad android:enableOnBackInvokedCallback del manifest.

JaimeGarcia-Jimmy committed:
-Quitar strings hardcodeadas y obtenerlas del archivo strings.xml
-Modificacion de archivos de UI, para separar textfields de etiquetas y valores.

JaimeGarcia-Jimmy committed:
-Se modificaron todos los RecyclerViews para utilizar ListAdapter.
